core/cache/resident.go

```diff
@@ -75,16 +87,80 @@ func (m *residentManager) new() *Resident {
 		nextResourceId: func() uint64 { return m.resourceCount.next() },
 		workers:        make(map[uint64]worker.Worker),
 	}
+
+	m.mu.Lock()
 	m.residents[r.id] = r
+	m.mu.Unlock()
+
 	return r
 }
 
-func (m *residentManager) evict(id uint64) {
-	// TODO (manadart 2019-04-17): TBC when the mark/sweep logic is added.
+// mark sets all of the manager's residents to be stale.
+// The boolean return indicates if there were any residents to mark.
+func (m *residentManager) mark() bool {
+	m.mu.Lock()
+	for _, r := range m.residents {
+		r.setStale(true)
+	}
+	m.mu.Unlock()
+
+	return len(m.residents) > 0
+}
+
+// sweep removes stale cache residents in descending order of ID.
+func (m *residentManager) sweep() <-chan struct{} {
+	removalIds, removalMessages := m.evictions()
+
+	finished := make(chan struct{})
+
+	go func() {
+	loop:
+		for _, id := range removalIds {
+			select {
+			case m.removals <- removalMessages[id]:
+			case <-m.dying:
```

Comment:
I'm happy to see that even if you get <-m.dying you still make sure to close finished. However, I would like to be defensive in the caller code as well (hence the concern about the bare read.)

Reply:
Threaded the dying channel there too.